tests/unit/test_pipeline_graph.py

```diff
@@ -307,3 +313,172 @@ def test_spark_is_required(default_config_params, requires_spark):
+    expected_nodes = {
+        "input_data",
+        "step_1_python_pandas",
+        "step_2_python_pandas",
+        "step_3_loop_1_step_3_python_pandas",
+        "step_3_4",
+        "results",
+    }
+    expected_edges = {
+        ("input_data", "step_1_python_pandas"): {
+            "input_slot_name": "step_1_main_input",
+            "output_slot_name": "all",
+            "validator": validate_input_file_dummy,
+            "env_var": "DUMMY_CONTAINER_MAIN_INPUT_FILE_PATHS",
+            "filepaths": (
+                Path(f"{test_dir}/input_data1/file1.csv"),
+                Path(f"{test_dir}/input_data2/file2.csv"),
+            ),
+        },
+        ("input_data", "step_3_4"): {
+            "input_slot_name": "step_4_secondary_input",
+            "output_slot_name": "all",
+            "validator": validate_input_file_dummy,
+            "env_var": "DUMMY_CONTAINER_SECONDARY_INPUT_FILE_PATHS",
+            "filepaths": (
+                Path(f"{test_dir}/input_data1/file1.csv"),
+                Path(f"{test_dir}/input_data2/file2.csv"),
+            ),
+        },
+        ("step_1_python_pandas", "step_2_python_pandas"): {
+            "input_slot_name": "step_2_main_input",
+            "output_slot_name": "step_1_main_output",
+            "validator": validate_input_file_dummy,
+            "env_var": "DUMMY_CONTAINER_MAIN_INPUT_FILE_PATHS",
+            "filepaths": (Path("intermediate/step_1_python_pandas/result.parquet"),),
+        },
+        ("step_2_python_pandas", "step_3_loop_1_step_3_python_pandas"): {
+            "input_slot_name": "step_3_main_input",
+            "output_slot_name": "step_2_main_output",
+            "validator": validate_input_file_dummy,
+            "env_var": "DUMMY_CONTAINER_MAIN_INPUT_FILE_PATHS",
+            "filepaths": (Path("intermediate/step_2_python_pandas/result.parquet"),),
+        },
+        ("step_3_loop_1_step_3_python_pandas", "step_3_4"): {
+            "input_slot_name": "step_3_main_input",
+            "output_slot_name": "step_3_main_output",
+            "validator": validate_input_file_dummy,
+            "env_var": "DUMMY_CONTAINER_MAIN_INPUT_FILE_PATHS",
+            "filepaths": (
+                Path("intermediate/step_3_loop_1_step_3_python_pandas/result.parquet"),
+            ),
+        },
+        ("step_3_4", "results"): {
+            "input_slot_name": "result",
+            "output_slot_name": "step_4_main_output",
+            "validator": validate_input_file_dummy,
+            "env_var": None,
+            "filepaths": (Path("intermediate/step_3_4/result.parquet"),),
+        },
+    }
+    check_nodes_and_edges(pipeline_graph, expected_nodes, expected_edges)
+
+
+def test_cycle_error(default_config_params) -> None:
+    config_params = default_config_params
+    config_params["pipeline"] = COMBINED_IMPLEMENTATION_CONFIGS["with_iteration_cycle"]
+    # Add a cycle
+    with pytest.raises(ValueError):
```

Comment:
Could make this a bit more specific to check that it's actually the right kind of error

Reply:
It's actually failing for a duplicate slot at the moment, let me add the correct match and mark it skip with a TODO.